native/java/org/jpype/manager/TypeManager.java

```diff
@@ -397,6 +397,12 @@ private ClassDescriptor createClass(Class<?> cls, boolean special)
 
   private ClassDescriptor createOrdinaryClass(Class<?> cls, boolean special, boolean bases)
   {
+    // Verify the class will be loadable prior to creating the class.
+    // If we fail to do this then the class may end up crashing later when the
+    // members get populated which could leave us in a bad state.
+    cls.getMethods();
```

Comment:
So this will raise a Java exception, if the methods dependencies are not resolvable?

Reply:
Yes that is Java's behavior though you don't really see that from the documentation.   Apparently Java does a lazy load of the methods without resolving them.  When you attempt to traverse the methods it generates a runtime exception.

https://docs.oracle.com/javase/8/docs/api/java/lang/Class.html#getMethods--

Unfortunately there is no way to traverse a class and find only those methods that are linked.   I really don't remember this behavior from earlier versions of Java.  Perhaps this was introduced in Java 9 when they did the module change.

Reply:
I should likely also add a call to ``getDeclaredMethods`` as there are methods that appear on the declared list and do not appear on the ``getMethods`` list.   So there may be another way to fail if the protected method is also unresolvable.   

There is also the mystery of why for one database driver it fails the first time and succeeds the second.   I am worried that we some major library will have a test class unresolvable and somehow we won't work with it.   Of course if that were the case than nothing that operates on reflection would work.  But this is really trending on being a Java bug.   It is not clear that Java ``getMethods`` is supposed to fail like this, but it does.

Reply:
Eventually the exception thrown in getMethods is an implementation detail we cannot rely upon? Did you check if later major versions of the JDK do declare the throw?
I absolutely agree upon calling getDeclaredMethods as well, as it will add more failure safety. In the past JPype also allowed to call private/protected methods, so resolving their dependencies as well should be a good thing. I wouldn't consider it good practice of a library, if its classes have private implementation testing methods. So I'm not sure, how common this pattern would be.

Reply:
Constructors could contain unresolved deps as well, right?

Reply:
Well so far based on my testing I have seen ClassNotDefError and ClassNotFoundException both of which would be runtime errors or worse.   So it is not really unexpected that they didn't get a throws declaration.   Throws declarations are usually reserved for checked exceptions though unchecked can declared for documentation.  I am guessing that the Java system wasn't really intended to have classes that did not fully resolve.   The problem here is that rather than returning a list that we could iterate over and skip unresolved methods, they defined these as an array so it is all or none.   Later caching operations to speed up the JVM left us in the state we have.

I agree that we could just check getDeclaredMethods, getDeclaredFields, and getDeclaredConstructors.  In theory because we always probe the super class first this should cover all the options.

I do have some question on how to resolve this.   One approach would be fall back to the base class rather than failing.   If some internal class like DrillImpl can't be loaded because it was supposed to be a private implementation class and method we are returning is "Driver" which we can do automatically then there is another option.   JPype misbehaves a bit in that it   automatically wraps to the most derived class even if that class should be unavailable due to the module system.   I use this feature often in my work, but there is no requirement that we must do that when a class we are importing wasn't meant for accessing.   I think it would be possible to return the super rather than dying.   But then you can get silent failures, so I am not sure if that is a good idea.

Reply:
I'd not skip unresolved methods, as API users would be confused of partially loaded classes (mismatching to documentation) -- a clear error message would be preferable then.

If wrapping super classes, instead of impl classes, that sounds like a good idea.
Why would we want to access things, which are not intended to be (of course it is cool and we can hack around)? I remember we had a similar discussion, when talking about respecting the visibility of private/public attributes.

Which silent failures could this cause? I can only think of abstract methods (just raising an exception as they should be overridden by children).

Reply:
Lets consider the concequences.  We have a return of type X but the actual type is Y.  We try to build a wrapper for type Y but get an exception so we instead return wrapper X.  In Java this isnt an issue as the static type is X and they would have to cast, but in Python there is only the dynamic type Y that is expected.

If X is ``Driver`` and Y is ``DrillImpl`` then all makes sense because X is a useable class.  Where is gets ugly is what is X is Object?  Now it gets ugly.  Does the user have to cast or do we go to the super type?  What if the supertype is Object as all the class has is an Interface?  What if it has multiple interfaces?

The problem is that Java public classes no longer mean public.  Many times implementation classes are public just so they can be created from the package due to the lack of a friend concept.  Thus we cant tell is Y is really a public class or a private class which is public due to module access.

Returning the static type all the time is not an option as operating on a list would be a pain and very unpythonic.

It seems like an error is the safest, but there is no way to avoid the error and just get the base wrapper.  Thus we end up stuck as bad jars become completely unusable.

This is a hard problem.  No option is particularly good.

Reply:
As far as abstract classes.  If you get the abstract class then you can access any member declared in the abstract even if not defined.  Those method still come up in getMethods and we always do virtual lookup.  It is no different than interfaces.   You  can test this by casting a wrapper to its abstract base class and using dir and calling the methods.   Last I checked that functionality all worked.